### PR TITLE
implement `single_line_let_else_max_width`

### PR DESCRIPTION
**Pull request by ytmimi**
This is a follow up PR to #5690

This allows users to configure the maximum length of a single line `let-else` statements. `let-else` statements that otherwise meet the requirements to be formatted on a single line ([as described by the style guide](https://github.com/rust-lang/rust/blob/master/src/doc/style-guide/src/statements.md#else-blocks-let-else-statements)) will have their divergent `else block` formatted over multiple lines if they exceed this length.

### CODE REVIEW

**Comment by calebcartwright** [comment]
Thanks for getting this up so quickly! We're only going to get this one shot at it, so will need to make sure this is both complete/perfect (as much as it can be) and stable which is a bit of an usual path for a net-new config opt.

Will try to review later tonight after work

**Comment by calebcartwright** [comment]
also @fee1-dead if you have the bandwidth an are interested in giving this a review your perspective would be greatly appreciated

File: src/config/config_type.rs
Comment (fee1-dead) [inline src/config/config_type.rs:419]: These config values seems to be a bit repetitive. Would be nice to do as a followup to extract this to a macro/function.

File: src/items.rs
Comment (fee1-dead) [inline src/items.rs:143]: This arithmetic can be a bit confusing to read. Is there similar functionality somewhere else that can be extracted to a common function? Also, can we test that the width configured corresponds to the actual width of the code exactly (probably not required for now)?

File: src/config/config_type.rs
Comment (ytmimi) [reply src/config/config_type.rs:419]: Agreed that it's something we should improve in a follow up PR

File: src/items.rs
Comment (ytmimi) [reply src/items.rs:143]: The only place I know where we clamp the width between two different widths is when rewriting doc comments, but even then it's slightly different: https://github.com/rust-lang/rustfmt/blob/9386b32f5a72bb276a97461d33aab415314201d8/src/comment.rs#L773-L776

At the point where we perform this check we've already written `let {pat} = {expr} else` into the buffer. And we're trying to see how much room we've got left in order to write the divergent `block` on one line. `shape.width` is going to vary based on how indented we are so we need to determine the available space based on the smaller of `shape.width` and `single_line_let_else_max_width`. If `available_space` hits zero then we know ahead of time that we wont be writing the block on one line.

Comment (fee1-dead) [reply src/items.rs:143]: Thanks. Could you add that as a comment?

Comment (ytmimi) [reply src/items.rs:143]: Always happy to add more tests. Are there specific test cases you think I should add? I think adding one with comments in the divergent block would be useful to show that those will always be multi-lined regardless of what you specify for `single_line_let_else_max_width`.

I also want to highlight that I tried to be thorough about the boundary tests when implementing `let-else`. Check out [`tests/target/let_else.rs`](https://github.com/rust-lang/rustfmt/blob/master/tests/target/let_else.rs) when you have a chance and let me know if you think I could expand on whats already there 😁

Comment (fee1-dead) [reply src/items.rs:143]: Ah. I didn't see that test file. Looks good then.

Comment (ytmimi) [reply src/items.rs:143]: Great! I'll add that comment a little later. I appreciate you taking the time to look this over. Please let me know if you've got any other thoughts on how to improve the PR

Comment (ytmimi) [reply src/items.rs]: @fee1-dead I added the comments that you asked for and I expanded on the test cases. Let me know if you'd like to see any other changes or have any other advice to improve the PR!

File: src/config/options.rs
Comment (fee1-dead) [inline src/config/options.rs:240]: leftover quotation mark?

```suggestion
    // always format the divergent `else block` over multiple lines.
```

As an aside, I think these should all be doc comments instead of normal comments.

File: src/items.rs
Comment (fee1-dead) [inline src/items.rs:143]:
```suggestion
                // same line. The available space varies based on indentation so we clamp the width
```

**Review by fee1-dead — APPROVED** [review]
LGTM. Had a few nits

File: src/config/options.rs
Comment (ytmimi) [reply src/config/options.rs:240]: Thanks for catching that. It was late when I wrote this!

File: src/items.rs
Comment (ytmimi) [reply src/items.rs:143]: 🙏🏼 Thanks for catching this too

File: Configurations.md
Comment (calebcartwright) [inline Configurations.md:2415]: let's remove the semicolon here.

in this context we want to demonstrate snippets that are eligible for single line based on all other rules, so that whether they are wrapped is solely dependent upon length and the value of the config option. the semis are load bearing though which means the snippets would always have to be wrapped regardless

```suggestion
        return
```

File: Configurations.md
Comment (calebcartwright) [inline Configurations.md:2420]: i assume the intent of this one is to reinforce the load-bearing nature of the semi relative to the first snippet that's able to be single lined due to width. i'm not sure it's necessary, but it obviously doesn't hurt.

however, if that's indeed the motivation then what would you think about having it immediately following the one line snippet? i feel like that association would pop a bit more if they were sequential

File: Configurations.md
Comment (calebcartwright) [inline Configurations.md:2433]: Same comment wrt the semi

```suggestion
        return
```

File: Configurations.md
Comment (calebcartwright) [inline Configurations.md:2438]: Same ordering suggestion as prior section

File: Configurations.md
Comment (calebcartwright) [inline Configurations.md:2452]: Same ordering suggestion as prior sections

File: src/config/config_type.rs
Comment (calebcartwright) [reply src/config/config_type.rs:419]: very much agreed. this is my work from long ago and i remember having the same thought while writing it, but wasn't sufficiently motivated to do the macro work.

not as part of this pr, but i think either a FIXME comment or an issue would be helpful to mark some TBD work on this, otherwise it's likely to remain forgotten unless/until someone has to touch this part of the code again

File: src/config/mod.rs
Comment (calebcartwright) [inline src/config/mod.rs:62]: nit but I don't think we want to capture the "block" word as part of the code fence:

```suggestion
        let-else statements. A value of zero means always format the divergent `else` block \
```

File: src/config/mod.rs
Comment (calebcartwright) [inline src/config/mod.rs:481]: Same comment as above

```suggestion
                `else` block over multiple lines.";
```

File: src/config/options.rs
Comment (calebcartwright) [inline src/config/options.rs:240]: Same theme as prior comments

```suggestion
    // always format the divergent `else` block over multiple lines.
```

File: src/items.rs
Comment (calebcartwright) [inline src/items.rs:165]: I concur with some of the earlier discussions on this that the high level theme of limit derivation, checking if a formatted version can fit within that limit, else format over multiple lines, feels like a potentially recurrent pattern that could be encapsulated in some manner that lends itself to reuse.

However, even if so, that's the type of internal change we could make afterwards with no impact on resultant behavior so I'd also agree that should be punted to a hypothetical future iteration.

File: tests/target/configs/single_line_let_else_max_width/50.rs
Comment (calebcartwright) [inline tests/target/configs/single_line_let_else_max_width/50.rs:23]: Could we include some additional cases at least here (with the val set to 50) that cover the other init expr scenarios?

E.g.

```rust
// Because the initializer expr is multi-lined the else is placed on it's own line.
    let Some(x) =
        some_really_really_really_really_really_really_really_really_really_long_name____H
    else {
        return;
    };
```

As well as the multilined init expr that starts on the same line as the assignment operator and ends with a sequence of one or more parens/brackets/braces?

**Review by calebcartwright — COMMENTED** [review]
Implementation LGTM, thank you! Few requested changes/additions to the docs and tests then should be good to go

File: Configurations.md
Comment (ytmimi) [reply Configurations.md:2433]: Unfortunately I can't remove the `;`. The `configuration_snippet_tests` would fail without it since rustfmt will add the `;` by default.

File: Configurations.md
Comment (ytmimi) [reply Configurations.md:2415]: Unfortunately I can't remove the ;. The configuration_snippet_tests would fail without it since rustfmt will add the ; by default.

File: Configurations.md
Comment (ytmimi) [reply Configurations.md:2420]: Yeah, that's why it's there. I'll move it.

File: src/items.rs
Comment (ytmimi) [reply src/items.rs:165]: great! appreciate you sharing your thoughts on that!

**Comment by ytmimi** [comment]
@calebcartwright let me know if these additional test cases are good. I'm happy to add any others you think are necessary for the new option.

File: tests/target/configs/single_line_let_else_max_width/100.rs
Comment (calebcartwright) [inline tests/target/configs/single_line_let_else_max_width/100.rs:29]: Glad we added this test, as I'm unsure whether this is desired behavior.

It would appear that the prescriptions in https://doc.rust-lang.org/nightly/style-guide/statements.html#else-blocks-let-else-statements do not explicitly cover the scenario where the statement is multilined prior to the the `else` block with an initializer expression that is single line.

I feel like one could read this:

> Otherwise, the else keyword and opening brace should be placed on the next line after the end of the initializer expression, and the else keyword should have the same indentation level as the let keyword.

As implying that if any of the prior scenarios aren't true, then break before the `else` keyword, but that is rather implicit. Thoughts? :thinking:

Comment (calebcartwright) [reply tests/target/configs/single_line_let_else_max_width/100.rs:29]: posted in zulip for style team as well https://rust-lang.zulipchat.com/#narrow/stream/346005-t-style/topic/let-else.20style.20guide.20inconsistency.3F/near/370115141

Comment (ytmimi) [reply tests/target/configs/single_line_let_else_max_width/100.rs:29]: I'll investigate this case and look into what's goin on. Now that you point it out I think I would have expected it to be formatted like:

```rust
    let Expr::Slice(ast::ExprSlice {
        lower,
        upper,
        step,
        range: _,
    }) = slice.as_ref()
    else {
        return;
    };
```

Comment (ytmimi) [reply tests/target/configs/single_line_let_else_max_width/100.rs:29]: The changes to produce the above formatting are pretty minimal. Let me know if I should go ahead and amend this PR once you hear back from t-style.

**Comment by calebcartwright** [comment]
I wouldn't be opposed to merging this as-is, and tackling any potential changes needed to address the [above scenario](https://github.com/rust-lang/rustfmt/pull/5790#discussion_r1243961627) in a separate PR.

My only slight reservation is that it may create some ambiguity for folks subscribed to the tracking issue (i.e. this PR and the associated issue are the last pending "big steps" articulated there), though I guess the same would be true for the blog post.

Lmk what you think; I'm probably leaning towards merging but not a strongly held position

**Comment by ytmimi** [comment]
updating the PR now

**Comment by calebcartwright** [comment]
> updating the PR now

in that case I guess we'll park this until there's direction one way or another from t-style :smile: 

I think there's enough here to justify having separate commits, though I also think the current 10 could probably be condensed into a smaller number. That's not strictly necessary, but wanted to share my observation in case you concur and would be interested in adjusting the commit history

**Comment by ytmimi** [comment]
Yeah, there's definitely room to reduce the number of commits. I'm happy to rebase! I just figured it was easier to have individual commits while working on this.

If we get word from t-style I can always remove the last commit. I'll hold off on making changes until it's clear how to move forward, though I guess we have discretion to make this change without an amendment to the guide, since this is the implied formatting from the style guide, correct?

**Comment by calebcartwright** [comment]
> though I guess we have discretion to make this change without an amendment to the guide, since this is the implied formatting from the style guide, correct?

Eh, I suspect this is something the Style team needs to explicitly respond to one way or the other, even if the position is "we defer to tools to interpret", though I suspect that's the least likely outcome

**Comment by calebcartwright** [comment]
> > though I guess we have discretion to make this change without an amendment to the guide, since this is the implied formatting from the style guide, correct?
> 
> Eh, I suspect this is something the Style team needs to explicitly respond to one way or the other, even if the position is "we defer to tools to interpret", though I suspect that's the least likely outcome

Seems wrapping is the expected behavior, and I anticipate an update to the rules to make this more explicit. As such I'm probably going to merge this later tonight, and move the blog post out of draft.

**Comment by ytmimi** [comment]
alright, I'll try to get the rebase in soon, just have a few more things I need to do for work

**Comment by calebcartwright** [comment]
> alright, I'll try to get the rebase in soon, just have a few more things I need to do for work

sorry I forgot about modifying the commit history. no rush, whenever you're ready.

i just wanted to share that we had an answer on the expected behavior from the style team

File: Configurations.md
Comment (ytmimi) [reply Configurations.md:2417]: I think the idea has briefly come up before. It's a change that I'd really like to see! I think it'll also be very helpful if we need to demonstrate formatting differences between `version` (and eventually `style_edition`)

**Comment by ytmimi** [comment]
@calebcartwright I think we'er good to go now. I've broken this up into 3 commits. The first fully implements `single_line_let_else_max_width`, while the other 2 make tweaks to the `let-else` formatting code.

Of those 2 extra commits the first inverts some existing logic, but doesn't change any formatting behavior, and the second implements wrapping the `else` when the pattern is too long.

**Pull request merged**
